src/framework/actions/combat.ts

```diff
@@ -0,0 +1,62 @@
+import { GameState, InstanceID } from "../types";
+import {
+  canAttack,
+  getAttack,
+  getHealth,
+  getPossibleAttackTargets,
+} from "../queries/combat";
+import { getInstance, getPlayer } from "../queries/common";
+
+export const attackInstance = (
+  $: GameState,
+  attackerId: InstanceID,
+  defenderId: InstanceID,
+): void => {
+  const attacker = getInstance($, attackerId);
+  const defender = getInstance($, defenderId);
+  if (!attacker) {
+    throw new Error(`id ${attackerId} is not an instance`);
+  }
+  if (!defender) {
+    throw new Error(`id ${defenderId} is not an instance`);
+  }
+
+  if (!canAttack($, attackerId)) {
+    throw new Error(`instance ${attackerId} can't attack`);
+  }
+
+  if (!getPossibleAttackTargets($, attackerId).includes(defenderId)) {
+    throw new Error(`instance ${attackerId} can't attack instance\
+ ${defenderId}`);
+  }
+
+  defender.damage += getAttack($, attackerId);
+  attacker.damage += getAttack($, defenderId);
+
+  if (getHealth($, attackerId) <= 0) {
+    killInstance($, attackerId);
+  }
+
+  if (getHealth($, defenderId) <= 0) {
+    killInstance($, defenderId);
+  }
+};
+
+const killInstance = (
+  $: GameState,
+  iid: InstanceID,
+): void => {
+  const I = getInstance($, iid);
+  if (!I) {
+    throw new Error(`id ${iid} is not an instance`);
+  }
+
+  if (I.card) {
```

Comment:
This is slightly wrong. `I.card` will always be set (it is used to hold the definition of the instance, even if it's not a standard deck card). To know what to do with the card, we need to do some logic. If it's a hero, then we wouldn't put it in the discard (and ofc need to track the summonability of that hero). If it's a token, it just fizzles. Any other card would go into the discard.